File: backend/app/schemas/author.py
"""作者端 Schema。"""

from pydantic import BaseModel, Field


class AuthorBookOut(BaseModel):
    id: int
    title: str
    description: str
    cover: str
    category: str
    status: str
    admin_delisted: bool = False
    word_count: int
    chapter_count: int
    updated_at: str


class CreateBookRequest(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=2000)
    category: str = Field(default="其他", max_length=50)
    cover: str | None = None


class UpdateBookRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=2000)
    category: str | None = None
    cover: str | None = None


class AuthorChapterOut(BaseModel):
    id: int
    title: str
    word_count: int
    sort_order: int
    updated_at: str


class AuthorChapterDetailOut(AuthorChapterOut):
    content: str


class CreateChapterRequest(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    content: str = Field(default="", max_length=50000)


class UpdateChapterRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=100)
    content: str | None = Field(default=None, max_length=50000)


class ReorderChaptersRequest(BaseModel):
    chapter_ids: list[int] = Field(min_length=1)

File: frontend/src/api/author.ts
import { request } from '@/api/request'

export interface AuthorBook {
  id: number
  title: string
  description: string
  cover: string
  category: string
  status: 'draft' | 'published' | 'unpublished'
  adminDelisted?: boolean
  wordCount: number
  chapterCount: number
  updatedAt: string
}

export function fetchAuthorBooks() {
  return request<AuthorBook[]>({ method: 'GET', url: '/author/books' })
}

export function createAuthorBook(data: {
  title: string
  description?: string
  category?: string
  cover?: string
}) {
  return request<AuthorBook>({ method: 'POST', url: '/author/books', data })
}

export function updateAuthorBook(
  bookId: number,
  data: { title?: string; description?: string; category?: string; cover?: string },
) {
  return request<AuthorBook>({ method: 'PATCH', url: `/author/books/${bookId}`, data })
}

export function publishAuthorBook(bookId: number) {
  return request<AuthorBook>({ method: 'POST', url: `/author/books/${bookId}/publish` })
}

export function unpublishAuthorBook(bookId: number) {
  return request<AuthorBook>({ method: 'POST', url: `/author/books/${bookId}/unpublish` })
}

export function deleteAuthorBook(bookId: number) {
  return request<null>({ method: 'DELETE', url: `/author/books/${bookId}` })
}

export interface AuthorChapter {
  id: number
  title: string
  wordCount: number
  sortOrder: number
  updatedAt: string
}

export function fetchAuthorChapters(bookId: number) {
  return request<AuthorChapter[]>({
    method: 'GET',
    url: `/author/books/${bookId}/chapters`,
  })
}

export function createAuthorChapter(
  bookId: number,
  data: { title: string; content?: string },
) {
  return request<AuthorChapter>({
    method: 'POST',
    url: `/author/books/${bookId}/chapters`,
    data,
  })
}

export function updateAuthorChapter(
  chapterId: number,
  data: { title?: string; content?: string },
) {
  return request<AuthorChapter>({
    method: 'PATCH',
    url: `/author/chapters/${chapterId}`,
    data,
  })
}

export function deleteAuthorChapter(chapterId: number) {
  return request<null>({ method: 'DELETE', url: `/author/chapters/${chapterId}` })
}

export function reorderAuthorChapters(bookId: number, chapterIds: number[]) {
  return request<AuthorChapter[]>({
    method: 'PUT',
    url: `/author/books/${bookId}/chapters/reorder`,
    data: { chapterIds },
  })
}

export function fetchAuthorChapterContent(chapterId: number) {
  return request<{ id: number; title: string; content: string; wordCount: number }>({
    method: 'GET',
    url: `/author/chapters/${chapterId}`,
  })
}
